python_modules/dagster/dagster/_model/__init__.py

```diff
@@ -58,3 +64,104 @@ def model_construct(cls, **kwargs: Any) -> Self:
             return super().model_construct(**kwargs)  # type: ignore
         else:
             return super().construct(**kwargs)
+
+
+T = TypeVar("T", bound=Type)
+
+
+def _banned(*args, **kwargs):
+    raise Exception("This method is not allowed on `@dagster_model`s.")
+
+
+def _dagster_model_transform(
+    cls: T,
+    *,
+    enable_cached_method: bool,
+) -> T:
+    field_set = {
+        **cls.__annotations__,
+        **({ALT_CACHED_METHOD_CACHE_FIELD: Any} if enable_cached_method else {}),
+    }
+    base = NamedTuple(f"_{cls.__name__}", field_set.items())
+    orig_new = base.__new__
+
+    checks = {
+        name: build_check_call(ttype=ttype, name=name)
+        for name, ttype in cls.__annotations__.items()
+    }
+
+    def __checked_new__(cls, *args, **kwargs):
+        for key, fn in checks.items():
+            fn(kwargs[key])
+
+        cache_fields = {ALT_CACHED_METHOD_CACHE_FIELD: {}} if enable_cached_method else {}
```

Comment:
to see this idea through would also need to
* make sure `copy` ignores it 
* make sure `serdes` ignores it